tests/ipc_shutdown.c

```diff
@@ -110,9 +111,9 @@ int main ()
     for (j = 0; j != TEST_LOOPS; ++j) {
         for (i = 0; i != TEST2_THREAD_COUNT; ++i)
             nn_thread_init (&threads [i], routine2, NULL);
-        active = TEST2_THREAD_COUNT;
+        nn_atomic_init(&active,TEST2_THREAD_COUNT);
 
-        while (active) {
+        while (nn_atomic_fetch(&active)) {
```

Comment:
Volatile was never adequate for multithreading.
So atomic_fetch is better, IMHO. Or Atomic_ flag for structure, but for C11

Reply:
The point is that a *read* against this is *cannot ever be zero* if other threads are running.  I.e. this can be lockless/atomic-free.  The volatile bit is necessary to just make sure we check memory again and don't rely on cache.

The *decrement* on the other hand, is prone to races, and does need the volatile, as you rightly notice.